### PR TITLE
Weighted sampler

### PR DESCRIPTION
**Pull request by dmylzenova**
This PR allows sampling points according to weights matrix, e.g. support correlation metrics.

### CODE REVIEW

File: seismiqb/src/horizon.py [hunk 1]
```diff
@@ -18,7 +18,6 @@
 from .plotters import plot_image
 
 
-
```
Comment (SergeyTsimfer) [inline seismiqb/src/horizon.py:21]: Please don't

File: seismiqb/src/horizon.py [hunk 2]
```diff
@@ -382,6 +381,7 @@ def __init__(self, storage, geometry, name=None, **kwargs):
         self._h_min, self._h_max = None, None
         self._h_mean, self._h_std = None, None
         self._horizon_metrics = None
+        self.support_corrs_matrix = None
```
Comment (SergeyTsimfer) [inline seismiqb/src/horizon.py:384]: `self._support_corrs`

File: seismiqb/src/horizon.py [hunk 3]
```diff
@@ -516,6 +516,8 @@ def reset_storage(self, storage=None):
         self._h_min, self._h_max = None, None
         self._h_mean, self._h_std = None, None
         self._len = None
+        self._horizon_metrics = None
```
Comment (SergeyTsimfer) [inline seismiqb/src/horizon.py:519]: I think it is time to remove the `horizon_metrics` property altogether

File: seismiqb/src/horizon.py [hunk 4]
```diff
@@ -1141,6 +1152,16 @@ def grad_along_axis(self, axis=0):
...
         return grad
 
+    @property
+    def support_corrs(self):
+        if self.support_corrs_matrix is None:
+            self.support_corrs_matrix = self._support_corrs()
+        return self.support_corrs_matrix
+
+
+    def _support_corrs(self, supports=50, **kwargs):
```
Comment (SergeyTsimfer) [inline seismiqb/src/horizon.py:1162]: As far as I can see, it is not needed in the current approach

File: seismiqb/src/horizon.py [hunk 5]
```diff
@@ -1165,8 +1186,9 @@ def evaluate(self, compute_metric=True, supports=50, plot=True, savepath=None, p
         """
         printer(dedent(msg))
         if compute_metric:
-            return self.horizon_metrics.evaluate('support_corrs', supports=supports, agg='nanmean',
-                                                 plot=plot, savepath=savepath, **kwargs)
+            self.support_corrs_matrix = self.horizon_metrics.evaluate('support_corrs', supports=supports, agg='nanmean',
```
Comment (SergeyTsimfer) [inline seismiqb/src/horizon.py:1189]: Move actual evaluation to the property

**Comment by roman-kh** [comment]
@SergeyTsimfer Closing?

**Pull request merged**